rfcs/2021-07-20-8288-csv-enrichment.md

```diff
@@ -0,0 +1,170 @@
+# RFC 8288 - 2021-07-20 - CSV Enrichment
+
+Vector needs to allow users to enrich the events flowing through the topology
+using through the topology using a CSV file. This RFC proposes adding a Vrl
+function that can perform this lookup.
+
+## Scope
+
+The RFC will cover a simple VRL function that will lookup a single row from a
+CSV file. 
+
+The lookup will be a simple sequential scan. The first version of this
+function will not handle any indexing of this data. Above a certain size
+indexing will provide substantial performance benefits, so it is something that
+we should definitely plan for implementing at a future stage. The RFC will 
+briefly discuss this, but will not go into any specific indexing strategies.
+
+## Motivation
+
+Users would like to enrich events flowing through Vector with extra data 
+provided by a CSV file. 
+
+## Doc-level Proposal
+
+Two remap functions:
+
+### `lookup_csv_row`
+
+This function will look up a single row within the dataset. If a single row is 
+found that data is returned as an object, otherwise this function will error. 
+
+### `lookup_csv_rows` 
+
+This function returns the results as an array of objects. If no row is found an
+empty array is returned. Multiple results just result in multiple rows in the 
+array.
+
+### Parameters 
+
+*filename*
+
+The filename is the path to the filename of the csv file to lookup. This
+filename has to be a static string (cannot be changed at runtime) so that 
+Vector can:
+
+1. Validate that the file exists.
+2. Load the file into memory.
+
+The filename will be relative to the current working directory that Vector is
+running in.
+
+The CSV file must start with a title row that gives a name to each of the 
+columns provided in the file.
+
+*key*
+
+The CSV file may be encrypted. This parameter specifies a key to use to decrypt
+the file.
+
+Is it safe to assume/require the file to be encrypted using AES, or should we
+support other algorithms as well?
+
+*criteria*
+
+`criteria` is a single level, key/value object that specifies the fields and
+values to lookup. The fields must all match (AND) for the row to be returned.
+
+*case_insensitive*
+
+By default the search will be case sensitive, but this can be changed by 
+passing `true` to this parameter.
+
+## Internal Proposal
+
+- Describe your change as if you were presenting it to the Vector team.
+- Use lists, examples, and code blocks for efficient reading.
+- Be specific!
+
+The majority of the changes will be provided by adding two new functions within
+`vrl/stdlib` and some shared lookup code.
+
+Since the CSV file is loaded at bootime, it is not essential that it is loaded 
+using async. This makes things much simpler as currently the VRL compilation 
+process is entirely synchronous. No code outside of the new function in the
+vrl stdlib should need changing.
+
+The entire data file will be loaded into memory, so all lookups will be 
+performed in memory.
+
+### Indexing 
```

Comment:
At a certain point of complexity in requirements, we may want to push people to sqlite too.

We _could_ even use an in-memory sqlite database and populate it from the CSV at start time. Users could configure indexes if we opt for configuration like what Jean has above:

```
files.myothercsv = {  decrypt = { ... }, path = "/opt/mycsvs/other.csv", index = ["foo", "bar", ["baz", "boop"]}
```